Sources/PackageCollections/Storage/PackageCollectionsStorage.swift

```diff
@@ -0,0 +1,326 @@
+        }
+    }
+
+    func remove(identifier: PackageCollectionsModel.PackageCollectionIdentifier,
+                callback: @escaping (Result<Void, Error>) -> Void) {
+        self.queue.async {
+            do {
+                let query = "DELETE FROM PACKAGES_COLLECTIONS WHERE key == ?;"
+                try self.executeStatement(query) { statement -> Void in
+                    let bindings: [SQLite.SQLiteValue] = [
+                        .string(identifier.databaseKey()),
+                    ]
+                    try statement.bind(bindings)
+                    try statement.step()
+                }
+                callback(.success(()))
+            } catch {
+                callback(.failure(error))
+            }
+        }
+    }
+
+    func get(identifier: PackageCollectionsModel.PackageCollectionIdentifier,
+             callback: @escaping (Result<PackageCollectionsModel.PackageCollection, Error>) -> Void) {
+        self.queue.async {
+            do {
+                let query = "SELECT value FROM PACKAGES_COLLECTIONS WHERE key == ? LIMIT 1;"
+                let collection = try self.executeStatement(query) { statement -> PackageCollectionsModel.PackageCollection in
+                    try statement.bind([.string(identifier.databaseKey())])
+
+                    let row = try statement.step()
+                    guard let data = row?.blob(at: 0) else {
+                        throw NotFoundError("\(identifier)")
+                    }
+
+                    let collection = try self.jsonDecoder.decode(PackageCollectionsModel.PackageCollection.self, from: data)
+                    return collection
+                }
+                callback(.success(collection))
+            } catch {
+                callback(.failure(error))
+            }
+        }
+    }
+
+    func list(identifiers: [PackageCollectionsModel.PackageCollectionIdentifier]? = nil,
+              callback: @escaping (Result<[PackageCollectionsModel.PackageCollection], Error>) -> Void) {
+        self.queue.async {
+            do {
+                var blobs = [Data]()
+                if let identifiers = identifiers {
+                    // TODO: consider running these in parallel
+                    var index = 0
+                    while index < identifiers.count {
+                        let slice = identifiers[index ..< min(index + Self.batchSize, identifiers.count)]
+                        let query = "SELECT value FROM PACKAGES_COLLECTIONS WHERE key in (\(slice.map { _ in "?" }.joined(separator: ",")));"
+                        try self.executeStatement(query) { statement in
+                            try statement.bind(slice.compactMap { .string($0.databaseKey()) })
+                            while let row = try statement.step() {
+                                blobs.append(row.blob(at: 0))
+                            }
+                        }
+                        index += Self.batchSize
+                    }
```

Comment:
@yim-lee took another approach here - batching to a well know batch size. this can be further optimized to run in parallel but its unlikely folks would have so many feeds so we can do that later.

WDYT?

Reply:
ok 👍